src/test/TestUtils/FileUtils.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.IO.MemoryMappedFiles;
+
+namespace Microsoft.DotNet.CoreSetup.Test.HostActivation
+{
+    public static class FileUtils
+    {
+        // See: https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm
+        static int[] ComputeKMP_FailureFunction(byte[] pattern)
+        {
+            int[] table = new int[pattern.Length];
+            if (pattern.Length >= 1)
+            {
+                table[0] = -1;
+            }
+            if (pattern.Length >= 2)
+            {
+                table[1] = 0;
+            }
+
+            int pos = 2;
+            int cnd = 0;
+            while (pos < pattern.Length)
+            {
+                if (pattern[pos - 1] == pattern[cnd])
+                {
+                    table[pos] = cnd + 1;
+                    cnd++;
+                    pos++;
+                }
+                else if (cnd > 0)
+                {
+                    cnd = table[cnd];
+                }
+                else
+                {
+                    table[pos] = 0;
+                    pos++;
+                }
+            }
+            return table;
+        }
+
+        // See: https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm
+        static unsafe int KMP_Search(byte[] pattern, byte* bytes, int length)
+        {
+            int m = 0;
+            int i = 0;
+            int[] table = ComputeKMP_FailureFunction(pattern);
+
+            while (m + i < length)
+            {
+                if (pattern[i] == bytes[m + i])
+                {
+                    if (i == pattern.Length - 1)
+                    {
+                        return m;
+                    }
+                    i++;
+                }
+                else
+                {
+                    if (table[i] > -1)
+                    {
+                        m = m + i - table[i];
+                        i = table[i];
+                    }
+                    else
+                    {
+                        m++;
+                        i = 0;
+                    }
+                }
+            }
+
+            return -1;
+        }
+
+        public static unsafe void SearchAndReplace(string filePath, byte[] search, byte[] replace, bool terminateWithNul)
```

Comment:
Isn't this going to break a signing? I am unsure how that works on Unix, but on Windows I think we could be making failing very annoying to figure out if loading fails. Does Azure enforce signatures during a test run?

Reply:
My take on this is:
* Yes - it does break signing, but this is only for our testing, I think (and hope) that it won't have an effect there
* If it works in CI and on dev machines, then I think it's good enough... at least for now
* I can see us intentionally disabling these tests for environments where signing matters. There are relatively few of these tests, vast majority of tests should not need this capability. At least right now.
* If we extend the usage of test-only behavior, for example the discussed potential of starting to mock parts of the product, then we would need to revisit this again.

And lastly I was not able to come up with another mechanism which is provably not introducing potential vectors for product behavior change which didn't exist before. The argument is that if you can modify the product binary to change the mark, you can also already modify it to do whatever you want anyway. This is what this change is about, without it, setting these test-only environment variables can be done by pretty much anything and it has profound effect on the product behavior.